> i am currently using python3.8

I dont know why these fail ... it's nothing I have done so they are broken by default on my version.

```
AIL: test_should_create_local_path (test.buildlog_test.BuildLogTest)
----------------------------------------------------------------------
Traceback (most recent call last):
  File "/home/kikass/gitclones/lbuild/test/buildlog_test.py", line 101, in test_should_create_local_path
    self.assertEqual("out1", o1.local_filename_out("test"))
AssertionError: 'out1' != '../../../../../test/out1'
- out1
+ ../../../../../test/out1
```
 * i did not touch anything here ... dunno why this fails (must have been in there already)
 * probably a failure due to changes in os.relpath

```
FAIL: test_should_generate_xml (test.buildlog_test.BuildLogTest)
----------------------------------------------------------------------
Traceback (most recent call last):
  File "/home/kikass/gitclones/lbuild/test/buildlog_test.py", line 82, in test_should_generate_xml
    self.assertEqual(b"""<?xml version='1.0' encoding='UTF-8'?>
AssertionError: b"<?x[130 chars]urce>m1/in1</source>\n    <destination>out1</d[170 chars]g>\n" != b"<?x[130 chars]urce>../../../../m1/in1</source>\n    <destina[194 chars]g>\n"
```
 * i did not touch anything here ... dunno why this fails (must have been in there already)
 * probably a failure due to changes in os.relpath


```
FAIL: test_should_build_archive_modules (test.parser_test.ParserTest)
----------------------------------------------------------------------
Traceback (most recent call last):
  File "/usr/local/lib/python3.8/dist-packages/testfixtures/utils.py", line 83, in patched
    return func(*args, **keywargs)
  File "/home/kikass/gitclones/lbuild/test/parser_test.py", line 392, in test_should_build_archive_modules
    self.assertTrue(os.path.isfile(path))
AssertionError: False is not true
```

*  i did not touch anything here ... dunno why this fails (must have been in there already)
* probably because `outpath = tempdir.path` does not properly create a temporary directory, so the `isFile(path) == True` fails later


```
FAIL: test_should_build_jinja_2_modules (test.parser_test.ParserTest)
----------------------------------------------------------------------
Traceback (most recent call last):
  File "/usr/local/lib/python3.8/dist-packages/testfixtures/utils.py", line 83, in patched
    return func(*args, **keywargs)
  File "/home/kikass/gitclones/lbuild/test/parser_test.py", line 413, in test_should_build_jinja_2_modules
    self.assertTrue(os.path.isfile(os.path.join(outpath, "src/module3.cpp")))
AssertionError: False is not true
```

 * same as above ... random tempdir not beeing there

```
======================================================================
FAIL: test_should_build_modules (test.parser_test.ParserTest)
----------------------------------------------------------------------
Traceback (most recent call last):
  File "/usr/local/lib/python3.8/dist-packages/testfixtures/utils.py", line 83, in patched
    return func(*args, **keywargs)
  File "/home/kikass/gitclones/lbuild/test/parser_test.py", line 291, in test_should_build_modules
    self.assertTrue(os.path.isfile(os.path.join(outpath, "src/other.cpp")))
AssertionError: False is not true
```

 * same as above

--- a/README.md
+++ b/README.md
@@ -14,8 +14,10 @@ You can [install *lbuild* via PyPi][pypi]: `pip install lbuild`
 
 Projects using *lbuild*:
 
-- [modm generates a HAL for hundreds of embedded devices][modm] using *lbuild*
+- [modm generates a HAL for thousands of embedded devices][modm] using *lbuild*
   and a data-driven code generation pipeline.
+- [Taproot: a friendly control library and framework for RoboMaster robots][taproot]
+  uses *lbuild*.
 - [OUTPOST - Open modUlar sofTware PlatfOrm for SpacecrafT][outpost] uses *lbuild*
   to assemble an execution platform targeted at embedded systems running mission
   critical software.
@@ -31,11 +33,11 @@ Consider this repository:
  $ lbuild discover
 Parser(lbuild)
 ╰── Repository(repo @ ../repo)
-    ├── EnumerationOption(option) = value in [value, special]
+    ├── Option(option) = value in [value, special]
     ├── Module(repo:module)
-    │   ├── BooleanOption(option) = True in [True, False]
+    │   ├── Option(option) = yes in [yes, no]
     │   ├── Module(repo:module:submodule)
-    │   │   ╰── EnumerationOption(option) = REQUIRED in [1, 2, 3, 4, 5]
+    │   │   ╰── Option(option) = REQUIRED in [1, 2, 3, 4, 5]
     │   ╰── Module(repo:module:submodule2)
     ╰── Module(modm:module2)
 ```
@@ -112,11 +114,11 @@ main.cpp        project.xml
 .
 ├── main.cpp
 ├── repo
-│   ├── module
-│   │   ├── static.hpp
-│   │   ├── template_1.cpp
-│   │   ├── template_2.cpp
-│   │   └── template_3.cpp
+│   ├── module
+│   │   ├── static.hpp
+│   │   ├── template_1.cpp
+│   │   ├── template_2.cpp
+│   │   └── template_3.cpp
 ```
 
 
@@ -136,7 +138,7 @@ Before you can build a project you need to provide a configuration.
  $ lbuild --repository ../modm/repo.lb discover
 Parser(lbuild)
 ╰── Repository(modm @ ../modm)   modm: a barebone embedded library generator
-    ╰── EnumerationOption(target) = REQUIRED in [at90can128, at90can32, at90can64, ...
+    ╰── Option(target) = REQUIRED in [at90can128, at90can32, at90can64, ...
 ```
 
 This gives you an overview of the repositories and their options. In this case
@@ -158,25 +160,25 @@ for this specific repository option:
  $ lbuild -r ../modm/repo.lb --option modm:target=stm32f407vgt discover
 Parser(lbuild)
 ╰── Repository(modm @ ../modm)   modm: a barebone embedded library generator
-    ├── EnumerationOption(target) = stm32f407vgt in [at90can128, at90can32, at90can64, ...]
+    ├── Option(target) = stm32f407vgt in [at90can128, at90can32, at90can64, ...]
     ├── Configuration(modm:disco-f407vg)
     ├── Module(modm:board)   Board Support Packages
     │   ╰── Module(modm:board:disco-f469ni)   STM32F469IDISCOVERY
     ├── Module(modm:build)   Build System Generators
     │   ├── PathOption(build.path) = build/parent-folder in [String]
-    │   ├── StringOption(project.name) = parent-folder in [String]
+    │   ├── Option(project.name) = parent-folder in [String]
     │   ╰── Module(modm:build:scons)  SCons Build Script Generator
-    │       ├── BooleanOption(info.build) = False in [True, False]
-    │       ╰── EnumerationOption(info.git) = Disabled in [Disabled, Info, Info+Status]
+    │       ├── Option(info.build) = no in [yes, no]
+    │       ╰── Option(info.git) = Disabled in [Disabled, Info, Info+Status]
     ├── Module(modm:platform)   Platform HAL
     │   ├── Module(modm:platform:can)   Controller Area Network (CAN)
     │   │   ╰── Module(modm:platform:can:1)   Instance 1
-    │   │       ├── NumericOption(buffer.rx) = 32 in [1 .. 32 .. 65534]
-    │   │       ╰── NumericOption(buffer.tx) = 32 in [1 .. 32 .. 65534]
+    │   │       ├── Option(buffer.rx) = 32 in [1 .. 32 .. 65534]
+    │   │       ╰── Option(buffer.tx) = 32 in [1 .. 32 .. 65534]
     │   ├── Module(modm:platform:core)   ARM Cortex-M Core
-    │   │   ├── EnumerationOption(allocator) = newlib in [block, newlib, tlsf]
-    │   │   ├── NumericOption(main_stack_size) = 3072 in [256 .. 3072 .. 65536]
-    │   │   ╰── EnumerationOption(vector_table_location) = rom in [ram, rom]
+    │   │   ├── Option(allocator) = newlib in [block, newlib, tlsf]
+    │   │   ├── Option(main_stack_size) = 3072 in [256 .. 3072 .. 65536]
+    │   │   ╰── Option(vector_table_location) = rom in [ram, rom]
 ```
 
 You can now discover all module options in more detail:
@@ -251,6 +253,8 @@ which *lbuild* will search for in the current working directory.
        exactly where the configuration file is stored in the repo.
        See also `repo.add_configuration(name, path)`. -->
   <extends>repo:name_of_config</extends>
+  <!-- You can declare the *where* the output should be generated, default is cwd -->
+  <outpath>generated/folder</outpath>
   <options>
     <!-- Options are treated as key-value pairs -->
     <option name="repo:repo_option_name">value</option>
@@ -304,11 +308,14 @@ global functions and classes for use in all files:
   to be explicit.
 - `repopath(path)`: remaps paths relative to the repository file. You should use
   this to reference paths that are not related to your module.
-- `FileReader(path)`: reads the contents of a file and turns it into a string.
 - `listify(obj)`: turns obj into a list, maps `None` to empty list.
+- `listrify(obj)`: turns obj into a list of strings, maps `None` to empty list.
+- `uniquify(obj)`: turns obj into a unique list, maps `None` to empty list.
+- `FileReader(path)`: reads the contents of a file and turns it into a string.
 - `{*}Option(...)`: classes for describing options, [see Options](#Options).
 - `{*}Query(...)`: classes for sharing code and data, [see Queries](#Queries).
 - `{*}Collector(...)`: classes for describing metadata sinks, [see Collectors](#Collectors).
+- `Alias(...)`: links to other nodes, [see Aliases](#Aliases).
 
 
 ### Repositories
@@ -1177,6 +1184,7 @@ it is therefore recommended to:
 
 
 [modm]: https://modm.io/how-modm-works
+[taproot]: https://github.com/uw-advanced-robotics/taproot
 [outpost]: https://github.com/DLR-RY/outpost-core
 [jinja2]: http://jinja.pocoo.org
 [python]: https://www.python.org

--- a/lbuild/api.py
+++ b/lbuild/api.py
@@ -25,7 +25,7 @@ from lbuild.logger import CallCounter
 
 class Builder:
 
-    def __init__(self, cwd=None, config=None, options=None, collectors=None):
+    def __init__(self, cwd=None, config=None, options=None, collectors=None, outpath=None):
         """
         Build instance to invoke lbuild methods.
 
@@ -38,6 +38,7 @@ class Builder:
                 "key=value" format.
             collectors -- List of collector values. Must be list of strings in a
                 "key=value" format.
+            outpath -- Path where the output will be generated. Defaults to cwd.
         """
         if cwd is None:
             if config is None:
@@ -75,6 +76,16 @@ class Builder:
         self.config = file_config
         self.config.add_commandline_options(listify(options))
         self.config.add_commandline_collectors(listify(collectors))
+
+        # 1. outpath is None: use cwd
+        self.outpath = self.cwd
+        # 2. outpath from config: use that
+        if self.config.outpath is not None:
+            self.outpath = self.config.outpath
+        # 3. outpath from command line overwrites it both
+        if outpath is not None:
+            self.outpath = outpath
+
         self.parser = Parser(self.config)
 
     def _load_repositories(self, repos=None):
@@ -123,20 +134,20 @@ class Builder:
         self.parser.validate_modules(build_modules, complete)
         return (build_modules, CallCounter.levels)
 
-    def build(self, outpath, modules=None, simulate=False, use_symlinks=False):
+    def build(self, modules=None, simulate=False, use_symlinks=False):
         """
         Build the given set of modules.
 
         Args:
-            outpath -- Path where the output will be generated.
             modules -- List of modules which should be built. This list
                 is combined with modules given in the configuration
                 files.
             simulate -- If set to True simulate the build process. In
                 that case no output will be generated.
+            use_symlinks -- symlink instead of copy non-templated files.
         """
         build_modules = self._filter_modules(modules)
-        buildlog = BuildLog(outpath)
+        buildlog = BuildLog(outpath=self.outpath, cwd=self.cwd)
         lbuild.environment.SYMLINK_ON_COPY = use_symlinks
         self.parser.build_modules(build_modules, buildlog, simulate=simulate)
         return buildlog

--- a/lbuild/buildlog.py
+++ b/lbuild/buildlog.py
@@ -77,9 +77,10 @@ class BuildLog:
     a specific file.
     """
 
-    def __init__(self, outpath=None):
+    def __init__(self, outpath=None, cwd=None):
         self._operations = collections.defaultdict(list)
         self._metadata = collections.defaultdict(lambda: collections.defaultdict(set))
+        self.cwd = os.getcwd() if cwd is None else abspath(cwd)
         self.outpath = os.getcwd() if outpath is None else abspath(outpath)
 
         self._build_files = {}

--- a/lbuild/config.py
+++ b/lbuild/config.py
@@ -32,6 +32,7 @@ class ConfigNode(anytree.AnyNode):
     def __init__(self, parent=None):
         anytree.AnyNode.__init__(self, parent)
 
+        self._outpath = None
         self._cachefolder = None
         self._extends = collections.defaultdict(list)
         self._vcs = []
@@ -63,6 +64,10 @@ class ConfigNode(anytree.AnyNode):
         return self._vcs
 
     @property
+    def outpath(self):
+        return self._outpath
+
+    @property
     def cachefolder(self):
         if self._cachefolder is None:
             return str(Path(self.filename).parent / DEFAULT_CACHE_FOLDER)
@@ -82,6 +87,8 @@ class ConfigNode(anytree.AnyNode):
         for node in list(self.siblings) + [self]:
             if node._cachefolder is not None:
                 config._cachefolder = node._cachefolder
+            if node._outpath is not None:
+                config._outpath = node._outpath
             config._extends.update(node._extends)
             config._vcs.extend(node._vcs)
             config._repositories.extend(node._repositories)
@@ -217,6 +224,11 @@ class ConfigNode(anytree.AnyNode):
 
         # Load all requested modules
         config._modules = xmltree.xpath('modules/module/text()')
+
+        # Load output path for lbuild and modules
+        outpath = xmltree.find("outpath")
+        if outpath is not None:
+            config._outpath = ConfigNode._rel_path(outpath.text, configpath)
 
         # Load options
         for option_node in xmltree.xpath('options/option'):

--- a/lbuild/environment.py
+++ b/lbuild/environment.py
@@ -94,7 +94,8 @@ class Environment:
         self.__module = module
         self.__modulepath = module._filepath
         self.__repopath = module.repository._filepath
-        self.__outpath = buildlog.outpath if buildlog else None
+        self.__outpath = buildlog.outpath if buildlog is not None else None
+        self.__cwdpath = buildlog.cwd if buildlog is not None else None
 
         self.buildlog = buildlog
         self.__template_environment = None
@@ -322,6 +323,20 @@ class Environment:
             return os.path.join(self.__outpath, *path)
 
         return os.path.join(self.__outpath, basepath, *path)
+
+    def cwdpath(self, *path):
+        """Relocate given path to current working directory"""
+        return os.path.join(self.__cwdpath, *path)
+
+    def cwdoutpath(self, *path):
+        """Relocate given path to the outpath in respect to current working directory"""
+        return self.outpath("", basepath="")
+
+    def relcwdoutpath(self, *path):
+        """Relocate given path to the outpath in respect to current working directory"""
+        relative = self.cwdpath("")
+        path = os.path.join(self.cwdoutpath(""), *path)
+        return os.path.relpath(os.path.abspath(path), os.path.abspath(relative))
 
     def reloutpath(self, path, relative=None):
         if relative is None:

--- a/lbuild/facade.py
+++ b/lbuild/facade.py
@@ -257,6 +257,15 @@ class EnvironmentValidateFacade:
     def localpath(self, *path):
         return self._env.modulepath(*path)
 
+    def cwdpath(self, *path):
+        return self._env.cwdpath(*path)
+
+    def cwdoutpath(self, *path):
+        return self._env.pcwdoutpath(*path)
+
+    def relcwdoutpath(self, *path):
+        return self._env.relcwdoutpath(*path)
+
     def __getitem__(self, key):
         return self._env.options[key]
 

--- a/lbuild/main.py
+++ b/lbuild/main.py
@@ -22,7 +22,7 @@ from lbuild.format import format_option_short_description
 
 from lbuild.api import Builder
 
-__version__ = '1.18.1'
+__version__ = '1.19.0'
 
 
 class InitAction:
@@ -316,7 +316,7 @@ class BuildAction(ManipulationActionBase):
 
     @staticmethod
     def perform(args, builder):
-        buildlog = builder.build(args.path, args.modules, simulate=args.simulate,
+        buildlog = builder.build(args.modules, simulate=args.simulate,
                                  use_symlinks=args.symlink)
 
         if args.simulate:
@@ -445,8 +445,8 @@ def prepare_argument_parser():
         '-p',
         '--path',
         dest='path',
-        default='.',
-        help="Path in which the library will be generated (default: '%(default)s').")
+        default=None,
+        help="Path in which the library will be generated (default: '.').")
     argument_parser.add_argument(
         '-D',
         '--option',
@@ -518,7 +518,7 @@ def run(args):
     except AttributeError:
         raise lbuild.exception.LbuildArgumentException("No command specified!")
 
-    builder = Builder(config=args.config, options=args.options, collectors=args.collectors)
+    builder = Builder(config=args.config, options=args.options, collectors=args.collectors, outpath=args.path)
     return command(args, builder)
 
 

--- a/lbuild/resources/configuration.xsd
+++ b/lbuild/resources/configuration.xsd
@@ -6,6 +6,7 @@
       <xsd:choice maxOccurs="unbounded">
         <xsd:element name="repositories" type="RepositoriesType" minOccurs="0" maxOccurs="1"/>
         <xsd:element name="extends" type="xsd:string" minOccurs="0" maxOccurs="unbounded"/>
+        <xsd:element name="outpath" type="xsd:string" minOccurs="0" maxOccurs="1"/>
         <xsd:element name="options" type="OptionsType" minOccurs="0" maxOccurs="1"/>
         <xsd:element name="modules" type="ModulesType" minOccurs="0" maxOccurs="1"/>
         <xsd:element name="collectors" type="CollectorsType" minOccurs="0" maxOccurs="1" />

--- a/test/api_test.py
+++ b/test/api_test.py
@@ -38,6 +38,7 @@ class ApiTest(unittest.TestCase):
             "repositories": list(),
             "vcs": list(),
             "cachefolder": ".lbuild_cache",
+            "outpath": None,
         }
         defaults.update(kw)
         for key, default in defaults.items():
@@ -51,11 +52,17 @@ class ApiTest(unittest.TestCase):
         api = lbuild.api.Builder(config="project.xml")
         self._assert_config(api)
 
+    def test_project_builder_outpath(self):
+        outputPath = "testpath"
+        api = lbuild.api.Builder(cwd=self._get_path("parser/api"), config="project.xml", outpath=outputPath)
+        self.assertEqual(api.outpath, outputPath)
+
     def test_config_builder(self):
         api = lbuild.api.Builder(config=self._get_path("parser/api/simple.xml"))
         self.assertEqual(api.cwd, self._get_path("parser/api"))
         self._assert_config(api,
             filename=self._rel_path("parser/api/simple.xml"),
+            outpath=self._get_path("parser/api/build/folder"),
             options={"repo:option": ("value", self._get_path("parser/api/simple.xml")),
                      "repo:module:option": ("value", self._get_path("parser/api/simple.xml"))},
             modules=["repo:module"],

--- a/test/config_test.py
+++ b/test/config_test.py
@@ -74,6 +74,7 @@ class ConfigTest(unittest.TestCase):
         config = self._parse_config("configfile/project.xml")
         del os.environ["LBUILD_TEST_REPOHOME"]
         self.assertEqual(self._get_path("configfile/hello_there"), abspath(config.cachefolder))
+        self.assertEqual(self._get_path("configfile/build/folder"), abspath(config.outpath))
 
         self.assertEqual(2, len(config.repositories))
         self.assertIn(self._get_path("repo2/repo2.lb"), config.repositories)
@@ -103,6 +104,7 @@ class ConfigTest(unittest.TestCase):
         self.assertEqual(config.collectors[1][1], 'value2')
         self.assertEqual(config.collectors[2][0], 'repo1:collect_empty')
         self.assertEqual(config.collectors[2][1], '')
+
 
     def test_should_parse_base_configuration(self):
         config = self._parse_config("configfile_inheritance/depth_0.xml")

--- a/test/resources/config/configfile/project.xml
+++ b/test/resources/config/configfile/project.xml
@@ -10,6 +10,8 @@
 		<cache>hello_there/</cache>
 	</repositories>
 
+	<outpath>build/folder</outpath>
+
 	<options>
 		<option name=":target">hosted</option>
 		<option name="repo1:foo">43</option>

--- a/test/resources/parser/api/simple.xml
+++ b/test/resources/parser/api/simple.xml
@@ -6,6 +6,8 @@
 		</repository>
 	</repositories>
 
+	<outpath>build/folder</outpath>
+
 	<options>
 		<option name="repo:option">value</option>
 		<option name="repo:module:option">value</option>

